Source/Details/ASCollectionGalleryLayoutDelegate.mm

```diff
@@ -105,6 +110,31 @@ + (ASCollectionLayoutState *)calculateLayoutWithContext:(ASCollectionLayoutConte
   NSMutableArray<_ASGalleryLayoutItem *> *children = ASArrayByFlatMapping(elements.itemElements,
                                                                           ASCollectionElement *element,
                                                                           [[_ASGalleryLayoutItem alloc] initWithItemSize:itemSize collectionElement:element]);
+  NSInteger numberOfSupplementaryelementsBefore = 0;;
+  if ([context.elements.supplementaryElementKinds count]) {
+    for (int sectionIndex = 0; sectionIndex< context.elements.numberOfSections; ++sectionIndex) {
+      for (NSString *kind in context.elements.supplementaryElementKinds) {
+        if ([kind isEqualToString:UICollectionElementKindSectionHeader]) {
+          ASCollectionElement *headerElement = [context.elements supplementaryElementOfKind:UICollectionElementKindSectionHeader atIndexPath:[NSIndexPath indexPathForRow:0 inSection:sectionIndex]];
+          if (headerElement) {
+            NSInteger index = sectionIndex > 0 ? [context.elements numberOfItemsInSection:sectionIndex - 1] : 0;
```

Comment:
I believe this `index` won't be valid for the 3rd section and below. The reason is that `numberOfItemsInSection` returns the number in the previous section, not the total number from the 1st section to the previous one.

I understand that you're trying to inject header and footer elements at the right positions after all item elements were generated above via `ASArrayByFlatMapping`. While this approach should work, I believe it's a bit over-complicated, which will affect readability and maintainability in the long term. How about dropping the call to `ASArrayByFlatMapping` completely and generate the children array from the ground up? That is, for each section, add headers first, then items and finally footers, and only then, move on to the next section.